### PR TITLE
[actor] Funnel correlation actor

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- change funnel correlation to actor pattern and support groups
*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*
- add group tests

### CODE REVIEW

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 1]
```diff
@@ -1,144 +1,189 @@
...
+            """
 
-        from posthog.api.person import PersonSerializer
+        return actor_join_subquery, actor_join_subquery_params
 
-        return (
-            PersonSerializer(people, many=True).data,
-            len(results) > cast(int, self._filter.correlation_person_limit) - 1,
-        )
+    def _get_group_filters(self):
```
Comment (EDsCODE) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:166]: did this instead of morphing the group filtering into the subquery directly. Open to other options

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 2]
```diff
@@ -1,144 +1,189 @@
...
+        assert isinstance(self._filter.correlation_person_entity, Entity)
 
         funnel_persons_query, funnel_persons_params = self._funnel_correlation.get_funnel_persons_cte()
 
+        prop_filters = self._filter.correlation_person_entity.properties
+        prop_query, prop_params = FunnelEventQuery(self._filter, self._team.pk)._get_props(prop_filters)
+
         conversion_filter = (
-            f'funnel_people.steps {"=" if self._filter.correlation_persons_converted else "<>"} target_step'
+            f'AND person.steps {"=" if self._filter.correlation_persons_converted else "<>"} target_step'
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:61]: This particular change is confusing for me - looking below, where is the `person` table coming from? Shouldn't this be funnel_people?

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 3]
```diff
@@ -1,144 +1,189 @@
...
-        person_query, person_query_params = ClickhousePersonQuery(
-            self._filter,
-            self._team.pk,
-            entity=Entity({"id": "person", "type": "events", "properties": self._filter.correlation_property_values}),
-        ).get_query()
+        event_join_query = self._funnel_correlation._get_events_join_query()
 
         query = f"""
             WITH
                 funnel_people as ({funnel_persons_query}),
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:70]: Q: Given the extensive refactoring, why not rename it?

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 4]
```diff
@@ -1,144 +1,189 @@
...
-                %(funnel_step_names)s as funnel_step_names
+                %(target_step)s AS target_step
             SELECT
-                DISTINCT person.person_id as person_id
-            FROM events AS event
-                {event_join_query}
-                AND event.event = %(target_event)s
-                {conversion_filter}
-                {prop_query}
+                DISTINCT funnel_people.person_id as person_id
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:127]: While funnel_people needs to return data as person_id as a kludge we should solve, no reason we shouldn't `AS actor_id` here.

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 5]
```diff
@@ -1,144 +1,189 @@
...
+                toDateTime(%(date_from)s) AS date_from,
+                %(target_step)s AS target_step,
+                %(funnel_step_names)s as funnel_step_names
             SELECT
-                DISTINCT funnel_people.person_id as person_id
-            FROM funnel_people
-            JOIN ({person_query}) person
-                ON person.id = funnel_people.person_id
-            WHERE {conversion_filter}
+                DISTINCT person.person_id as person_id
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:76]: While funnel_people needs to return data as person_id as a kludge we should solve, no reason we shouldn't `AS actor_id` here.

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 6]
```diff
@@ -1,144 +1,189 @@
...
             else ""
         )
 
-        event_join_query = self._funnel_correlation._get_events_join_query()
+        actor_join_subquery, actor_join_subquery_params = self._get_actor_subquery()
+        group_filters, group_filters_params = self._get_group_filters()
 
         query = f"""
             WITH
                 funnel_people as ({funnel_persons_query}),
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:124]: Why not name this funnel_actors now for clarity?

File: ee/clickhouse/queries/funnels/funnel_correlation_persons.py [hunk 7]
```diff
@@ -1,144 +1,189 @@
...
         if not self._filter.correlation_person_limit:
             self._filter = self._filter.with_data({FUNNEL_CORRELATION_PERSON_LIMIT: 100})
 
-    def run(self):
-        """
-        Returns `ReturnList` type, generated by `serializers.serialize`, which returns the Person model.
-        """
+    def get_actors(
+        self,
+    ) -> Tuple[Union[QuerySet[Person], QuerySet[Group]], Union[List[SerializedGroup], List[SerializedPerson]]]:
```
Comment (macobo) [inline ee/clickhouse/queries/funnels/funnel_correlation_persons.py:32]: Nit: This return type is quite complex. There's IMO two clearer ways to write this:

```
Union[Tuple[QuerySet[Person], List[SerializedPerson]], Tuple[QuerySet[Group],List[SerializedGroup]]
```

Or using union types for Person | Group, SerializedPerson | SerializedGroup

**Review by macobo — COMMENTED** [review]
Will take a closer look later. In general this looks OK, though how one query works I have no clue.

Neil took a shortcut implementing groups support here in _not_ renaming things. However this seems to be making everything more complicated, not less so. I suggest doing the renaming in SQL where possible - stuff is really hard to follow because of it right now.

**Comment by neilkakkar** [comment]
Just a quick flyby: I was constrained by the funnel persons query earlier (that was still passing person_id vs actor_id / aggregation_target ) because other queries (like paths) depended on it, and I didn't want to do the full refactor for everything in the PRs I did.

But now, since everything else has(?) moved over to actor IDs, a lot of this can be simplified to use the same names!

**Review by macobo — APPROVED** [review]
Didn't go super deep but to unblock you I think this looks much clearer now! Solid work :muscle:
